src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/LspFileBasedProgramWorkspace.cs

```diff
@@ -0,0 +1,323 @@
+using Microsoft.CodeAnalysis.MSBuild;
+using Microsoft.CodeAnalysis.ProjectSystem;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.CodeAnalysis.Threading;
+using Microsoft.CodeAnalysis.Workspaces.ProjectSystem;
+using Microsoft.CommonLanguageServerProtocol.Framework;
+using Microsoft.Extensions.Logging;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.LanguageServer;
+
+/// <summary>
+/// Workspace which handles file-based programs ('dotnet run app.cs' scenario).
+/// </summary>
+internal sealed class LspFileBasedProgramWorkspace : Workspace, ILspService, ILspWorkspace, ILspFileBasedProgramWorkspace
+{
+    /// <summary>
+    /// A single gate for code that is adding work to <see cref="_projectsToLoadAndReload" />. This is just we don't have code simultaneously trying to load and unload solutions at once.
+    /// </summary>
+    private readonly SemaphoreSlim _gate = new SemaphoreSlim(initialCount: 1);
+
+    // TODO: should we use the LanguageServerProjectSystem here instead?
+    private readonly AsyncBatchingWorkQueue<(Uri uri, SourceText documentText, string languageId)> _projectsToLoadAndReload; // TODO: do we need to introduce a shutdown cancellation token for this?
+
+    private readonly ConcurrentDictionary<string, List<LoadedProject>> _loadedProjects = [];
+    private readonly ILspServices _lspServices;
+    private readonly IFileChangeWatcher _fileChangeWatcher;
+    private readonly IEnumerable<Lazy<IDynamicFileInfoProvider, FileExtensionsMetadata>> _dynamicFileInfoProviders;
+    private readonly ProjectTargetFrameworkManager _projectTargetFrameworkManager;
+    private readonly ILoggerFactory _loggerFactory;
+
+    public bool SupportsMutation => true;
+
+    // ILspLogger rationalize
+    ILogger iLogger;
+
+    public LspFileBasedProgramWorkspace(
+        ILspServices lspServices,
+        IFileChangeWatcher fileChangeWatcher,
+        [ImportMany] IEnumerable<Lazy<IDynamicFileInfoProvider, FileExtensionsMetadata>> dynamicFileInfoProviders,
+        ProjectTargetFrameworkManager projectTargetFrameworkManager,
+        ILoggerFactory loggerFactory,
+        IAsynchronousOperationListenerProvider listenerProvider,
+        HostServices hostServices) : base(hostServices, WorkspaceKind.FileBasedPrograms)
+    {
+        _lspServices = lspServices;
+        _fileChangeWatcher = fileChangeWatcher;
+        _dynamicFileInfoProviders = dynamicFileInfoProviders;
+        _projectTargetFrameworkManager = projectTargetFrameworkManager;
+        _loggerFactory = loggerFactory;
+        _projectsToLoadAndReload = new AsyncBatchingWorkQueue<(Uri uri, SourceText documentText, string languageId)>(
+            TimeSpan.FromMilliseconds(100),
+            LoadOrReloadProjectsAsync,
+            equalityComparer: null,
+            listenerProvider.GetListener(FeatureAttribute.Workspace),
+            CancellationToken.None);
```

Comment:
needs cancellation token.  You're a mef component, so you can import IThreadingContext and pass its disposal token here.  

Should not need the overload that takes an equality comaprer since you're passing null along.

Reply:
This is in the language server, so no ITheadingContext exists. In the place where Riki commented this from, there's a TODO asking if we should have one.